# 多元线性回归/对数线性回归（二选一）

## 一、多元线性回归
这部分的内容是要求大家完成多元线性回归，我们会先带着大家使用sklearn做一元线性回归的十折交叉验证，多元线性回归大家可以仿照着完成

### 1. 读取数据

In [1]:
import pandas as pd

In [2]:
# 读取数据
data = pd.read_csv('advertising.csv')


In [3]:
data.head()

,TV,Radio,Newspaper,Sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,12.0
3,151.5,41.3,58.5,16.5
4,180.8,10.8,58.4,17.9


### 2. 引入模型

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_predict

### 3. 使用sklearn完成一元线性回归的十折交叉验证验证

#### 创建模型

In [5]:
model = LinearRegression()

#### 选取数据

In [6]:
features = ['TV']
x = data[features]
y = data['Sales']

#### 做十折交叉验证的预测

In [7]:
prediction = cross_val_predict(model, x, y, cv = 10)

这十折交叉验证是按顺序做的，会先将前10%的数据作为测试集，然后会往后顺延到10%到20%，最后将这十份的预测值按顺序拼接后返回

In [8]:
prediction.shape

(200,)

### 4. 计算评价指标

#### MAE

In [9]:
mean_absolute_error(prediction, data['Sales'])

1.8484560436845014

#### RMSE

In [10]:
mean_squared_error(prediction, data['Sales']) ** 0.5

2.3118916484389405

### 5. 请你选择多种特征进行组合，完成多元线性回归，并对比不同的特征组合，它们训练出的模型在十折交叉验证上MAE与RMSE的差别，至少完成3组

###### 扩展：多项式回归（一元线性回归的扩展），尝试对部分特征进行变换，如将其二次幂，三次幂作为特征输入模型，观察模型在预测能力上的变化
###### 提示：多元线性回归，只要在上方的features这个list中，加入其他特征的名字就可以

In [11]:
# 定义要比较的特征组合
feature_combinations = [
    ['TV', 'Radio', 'Newspaper'],
    ['TV', 'Radio'],
    ['TV', 'Newspaper'],
    ['Radio', 'Newspaper'],
]

for features in feature_combinations:
    x = data[features]
    # 十折交叉验证预测
    prediction = cross_val_predict(model, x, y, cv = 10)
    mae = mean_absolute_error(prediction, y)
    rmse = mean_squared_error(prediction, y) ** 0.5
    print(features)
    print('MAE:', mae)
    print('RMSE:', rmse)
    print()


['TV', 'Radio', 'Newspaper']
MAE: 1.264454180776078
RMSE: 1.6857740069140652

['TV', 'Radio']
MAE: 1.2621936422519533
RMSE: 1.68189905182132

['TV', 'Newspaper']
MAE: 1.776707605359968
RMSE: 2.2492610419402657

['Radio', 'Newspaper']
MAE: 4.3163332911165995
RMSE: 5.042150499491994



###### 双击此处填写
1. 模型1使用的特征：TV、Radio、Newspaper
2. 模型2使用的特征：TV、Radio
3. 模型3使用的特征：TV、Newspaper
4. 模型4使用的特征：Radio、Newspaper

模型|MAE|RMSE
-|-|-
模型1 | 1.2645 | 1.6858
模型2 | 1.2622 | 1.6819
模型3 | 1.7767 | 2.2493
模型4 | 4.3163 | 5.0422


In [12]:
# 选做：多项式回归
data_poly = data.copy()
data_poly['TV2'] = data_poly['TV'] ** 2
data_poly['TV3'] = data_poly['TV'] ** 3

# 对比：仅TV、TV+TV^2、TV+TV^2+TV^3
poly_combinations = [
    ['TV'],
    ['TV', 'TV2'],
    ['TV', 'TV2', 'TV3'],
]

for features in poly_combinations:
    x = data_poly[features]
    prediction = cross_val_predict(model, x, y, cv = 10)
    mae = mean_absolute_error(prediction, y)
    rmse = mean_squared_error(prediction, y) ** 0.5
    print(features)
    print('MAE:', mae)
    print('RMSE:', rmse)
    print()

# 再尝试把 TV、Radio 的二次项加入多元模型
data_poly['Radio2'] = data_poly['Radio'] ** 2
x = data_poly[['TV', 'Radio', 'TV2', 'Radio2']]
prediction = cross_val_predict(model, x, y, cv = 10)
print(['TV', 'Radio', 'TV2', 'Radio2'])
print('MAE:', mean_absolute_error(prediction, y))
print('RMSE:', mean_squared_error(prediction, y) ** 0.5)


['TV']
MAE: 1.8484560436845014
RMSE: 2.3118916484389405

['TV', 'TV2']
MAE: 1.800279272444673
RMSE: 2.2571888068471573

['TV', 'TV2', 'TV3']
MAE: 1.8125967310188227
RMSE: 2.27376234629439

['TV', 'Radio', 'TV2', 'Radio2']
MAE: 1.1294911631581686
RMSE: 1.4915893318074538
